src/NLog/Targets/WebServiceTarget.cs

```diff
@@ -226,74 +227,230 @@ protected override void DoInvoke(object[] parameters, AsyncContinuation continua
 
-                            if (ex.MustBeRethrown())
-                            {
-                                throw;
+                                postPayload.Dispose();
+
+                                continuation(ex);
+                                if (flushCallback.Value != null)
+                                    flushCallback.Value(ex);
                             }
+                        });
+                }
+                catch (Exception ex)
+                {
+                    InternalLogger.Error(ex, "Error when sending to Webservice: {0}", this.Name);
+                    if (ex.MustBeRethrown())
+                    {
+                        throw;
+                    }
 
-                            continuation(ex);
-                        }
-                    });
+                    KeyValuePair<DateTime, AsyncContinuation> flushCallback;
+                    lock (this.SyncRoot)
+                    {
+                        if (_pendingRequests.TryGetValue(request, out flushCallback))
+                            _pendingRequests.Remove(request);
+                    }
+                    continuation(ex);
+                    if (flushCallback.Value != null)
+                        flushCallback.Value(ex);
+                }
             }
             else
             {
+                lock (this.SyncRoot)
+                    _pendingRequests.Add(request, new KeyValuePair<DateTime, AsyncContinuation>(DateTime.UtcNow, null));
                 sendContinuation(null);
             }
         }
 
         /// <summary>
+        /// Flush any pending log messages asynchronously (in case of asynchronous targets).
+        /// </summary>
+        /// <param name="asyncContinuation">The asynchronous continuation.</param>
+        protected override void FlushAsync(AsyncContinuation asyncContinuation)
+        {
+            // We want to be notified when, all pending requests have completed
+            if (_pendingRequests.Count == 0)
+            {
+                asyncContinuation(null);
+            }
+            else
+            {
+                int remainingCount = _pendingRequests.Count;
+                AsyncContinuation continuation =
+                    ex =>
+                    {
+                    // ignore exception
+                    if (System.Threading.Interlocked.Decrement(ref remainingCount) == 0)
+                        {
+                            asyncContinuation(null);
+                        }
+                    };
+
+                DateTime utcNowTimeout = DateTime.UtcNow.AddSeconds(-120);    // Default HttpWebRequest.Timeout = 100 secs
+
+                bool foundOldWebRequest = false;
+
+                foreach (var key in new List<HttpWebRequest>(_pendingRequests.Keys))
+                {
+                    KeyValuePair<DateTime, AsyncContinuation> flushCallback;
+                    if (_pendingRequests.TryGetValue(key, out flushCallback))
```

Comment:
FlushAsync is already protected by this.SyncRoot. Same lock-object is used for the other location.